server/src/main/java/io/druid/server/coordinator/DruidCoordinator.java

```diff
@@ -239,38 +241,40 @@ public boolean isLeader()
     }
 
     final DateTime now = new DateTime();
-    for (DataSegment segment : getAvailableDataSegments()) {
-      List<Rule> rules = metadataRuleManager.getRulesWithDefault(segment.getDataSource());
-      for (Rule rule : rules) {
-        if (rule instanceof LoadRule && rule.appliesTo(segment, now)) {
-          for (Map.Entry<String, Integer> entry : ((LoadRule) rule).getTieredReplicants()
-                                                                   .entrySet()) {
-            Object2LongOpenHashMap<String> dataSourceMap = retVal.get(entry.getKey());
-            if (dataSourceMap == null) {
-              dataSourceMap = new Object2LongOpenHashMap<>();
-              retVal.put(entry.getKey(), dataSourceMap);
-            }
 
-            int diff = Math.max(
-                entry.getValue() -
-                segmentReplicantLookup.getTotalReplicants(
-                    segment.getIdentifier(),
-                    entry.getKey()
-                ),
-                0
-            );
-            dataSourceMap.addTo(segment.getDataSource(), diff);
-          }
-          break;
-        }
-      }
+    for (final DataSegment segment : getAvailableDataSegments()) {
+      final List<Rule> rules = metadataRuleManager.getRulesWithDefault(segment.getDataSource());
+
+      final Optional<Rule> rule = rules
+          .stream()
+          .filter((final Rule r) -> r instanceof LoadRule && r.appliesTo(segment, now))
+          .findFirst();
+
+      rule.map((final Rule r) -> (LoadRule) r)
```

Comment:
IMO this monadic code with optional is Scala/Haskell like and it's harder to read than the former old-style Java code